_source/_posts/2022-05-16-jhipster-k8s-azure.md

```diff
@@ -0,0 +1,733 @@
+  value: prod,k8s,oauth2
+```
+
+Also in `k8s/registry-k8s/jhipster-registry.yml`, update the `ENCRYPT_KEY` value to use the same encryption key you used above in the Docker Compose section in the `.env` file.
+
+```bash
+- name: ENCRYPT_KEY
+  value: <your-encryption-key>
+```
+
+To configure the `store` service to use the Cosmo database, you need to put your connection string in `k8s/store-k8s/store-deployment.yml`.
+
+```bash
+- name: SPRING_DATA_MONGODB_URI
+  value: "<your-connection-string>"
+```
+
+Both the encryption key and the database connection string are sensitive values that need to be encrypted. You'll see how to do that just a little later in this tutorial.
+
+## Build Docker Images and Push to Docker Hub
+
+Previously you built the docker images, but you left them in the local repository. You need to upload them to Docker Hub so that Azure AKS can find them. If you haven't already [signed up for a Docker Hub account](https://hub.docker.com/), please do so now. 
+
+In each of the three directories (`blog`, `store`, and `gateway`), run the following command. Save your docker repo name in a Bash variable as shown below and you can copy and paste the commands and run them in each service directory.
+
+```bash
+DOCKER_REPO_NAME=<docker-repo-name>
+# in blog
+./gradlew bootJar -Pprod jib -Djib.to.image=$DOCKER_REPO_NAME/blog
+# in store
+./gradlew bootJar -Pprod jib -Djib.to.image=$DOCKER_REPO_NAME/store
+# in gateway
+./gradlew bootJar -Pprod jib -Djib.to.image=$DOCKER_REPO_NAME/gateway
+```
+
+To briefly explain what's happening here, take a look at the blog service's Kubernetes descriptor file. It defines a container named `blog-app` that uses the docker image `andrewcarterhughes/blog`, which is my docker repository name and the blog image.
+
+`k8s/blog-k8s/blog-deployment.yml`
+
+```yaml
+containers:
+  - name: blog-app
+    image: andrewcarterhughes/blog
+```
+
+Thus in the `k8s` directory, each service (store, blog, gateway, and registry) defines a container and a docker image to be run in that container, along with a whole lot of configuration (which is really a lot of what JHipster is bootstrapping for you). The registry does not have a project folder because it uses a stock image that can be pulled directly from the JHipster docker repository.
+
+One nice feature of Kubernetes is the ability to define [init containers](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/). These are containers that run before the main container and can be used to create or wait for necessary resources like databases. I noticed while I was debugging things in this app that a lot of the errors happened in the init containers. It's helpful to know this because if you try and inspect the main container log nothing will be there because the container hasn't even started you. You have to check the log for the init container that failed. The Kubernetes management tools that I mention below really come in handy for this.
```

Comment:
```suggestion
One nice feature of Kubernetes is the ability to define [init containers](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/). These are containers that run before the main container and can be used to create or wait for necessary resources like databases. I noticed while I was debugging things in this app that a lot of the errors happened in the init containers. It's helpful to know this because if you try and inspect the main container log nothing will be there because the container hasn't even started yet. You have to check the log for the init container that failed. The Kubernetes management tools that I mention below really come in handy for this.
```
typo - you-> yet ??